src/bin/pg_autoctl/env_utils.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * src/bin/pg_autoctl/env_utils.c
+ *   Utility functions for interacting with environment settings.
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "defaults.h"
+#include "env_utils.h"
+#include "log.h"
+
+
+/*
+ * env_empty returns true if the passed environment variable is the empty
+ * string. It returns false when the environment variable is not set or if it
+ * set but is something else than the empty string.
+ */
+bool
+env_empty(const char *name)
+{
+	char *envvalue = NULL;
+	if (name == NULL || strlen(name) == 0)
+	{
+		log_error("Failed to get environment setting. NULL or empty variable name is provided");
+		return false;
+	}
+
+	/*
+	 * Explanation of IGNORE-BANNED
+	 * getenv is safe here because we never provide null argument,
+	 * and only check the value it's length.
+	 */
+	envvalue = getenv(name); /* IGNORE-BANNED */
+	return envvalue != NULL && strlen(envvalue) == 0;
+
+}
+
+
+/*
+ * env_exists returns true if the passed environment variable exists in the
+ * environment, otherwise it returns false.
+ */
+bool
+env_exists(const char *name)
+{
+	if (name == NULL || strlen(name) == 0)
+	{
+		log_error("Failed to get environment setting. NULL or empty variable name is provided");
+		return false;
+	}
+
+	/*
+	 * Explanation of IGNORE-BANNED
+	 * getenv is safe here because we never provide null argument,
+	 * and only check if it returns NULL.
+	 */
+	return getenv(name) != NULL; /* IGNORE-BANNED */
+}
+
+
+/*
+ * get_env_copy_with_fallback copies the environment variable with "name" into
+ * the result buffer. It returns false when it fails. If the environment
+ * variable is not set the fallback string will be written in the buffer.
+ * Except when fallback is NULL, in that case an error is returned.
+ */
+bool
+get_env_copy_with_fallback(const char *name, char *result, int maxLength,
+						   const char* fallback)
+{
+	const char *envvalue = NULL;
+	size_t actualLength = 0;
+
+	if (name == NULL || strlen(name) == 0)
+	{
+		log_error("Failed to get environment setting. NULL or empty variable name is provided");
+		return false;
+	}
+
+	if (result == NULL)
+	{
+		log_error("Failed to get environment setting. Tried to store in NULL pointer");
+		return false;
+	}
+
+	/*
+	 * Explanation of IGNORE-BANNED
+	 * getenv is safe here because we never provide null argument,
+	 * and copy out the result immediately.
+	 */
+	envvalue = getenv(name); /* IGNORE-BANNED */
+	if (envvalue == NULL)
+	{
+		envvalue = fallback;
+		if (envvalue == NULL)
+		{
+			log_error("Failed to get %s environment setting. It was not set", name);
```

Comment:
I think we can improve this error message to “Failed to get value for environment variable \"%s\", which is  unset”.